### ISSUE TITLE
Weird index not found error on ArangoDB 3.5.3, when running queries, db server needs to be restarted

### ISSUE THREAD

**Issue by rocketraman** [arangodb/arangodb#11026]
I have the following error from ArangoDB 3.5.3, upon trying to run a query which seems to be stuck running forever. The query is showing in the Slow Queries log with a runtime of 169,650 seconds (so far). The only errors I see in my application logs are:

```
Caused by: com.arangodb.ArangoDBException: Response: 500, Error: 1590 - AQL: error, cluster node: 'PRMR-e6e30ee8-02f5-4707-8361-89cbc0f69c69', endpoint: 'tcp://arangodb-db-2.arangodb-dbs.default.svc.cluster.local:8529', error: 'Could not find index '214730889' in collection 's3010751'.' (while optimizing plan)
at com.arangodb.async.internal.velocystream.VstCommunicationAsync.lambda$execute$0(VstCommunicationAsync.java:66) ~[arangodb-java-driver-6.4.1.jar:?]
at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073) ~[?:?]
at com.arangodb.async.internal.velocystream.VstConnectionAsync.lambda$write$0(VstConnectionAsync.java:48) ~[arangodb-java-driver-6.4.1.jar:?]
at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
at com.arangodb.internal.velocystream.internal.MessageStore.consume(MessageStore.java:63) ~[arangodb-java-driver-6.4.1.jar:?]
at com.arangodb.internal.velocystream.internal.ChunkStore.checkCompleteness(ChunkStore.java:66) ~[arangodb-java-driver-6.4.1.jar:?]
```

Index '214730889' is indeed an index used by that query, but it definitely exists, so that message makes no sense whatsoever.

Note that in my case the query in question was a graph traversal relative to a specific object specified by `_key` that did `ANY` instead of `INBOUND` so it seems like it was scanning a massive dataset. Changing it to `INBOUND` and the query completes in less than half a second. However, this appears to be a problem in the graph query execution plan, because if the query just looked at the edges related to that vertex, it could have discarded the non-existing `OUTBOUND` connections immediately, so it should have been trivial extra work for this query execution to work with `ANY`.

Finally, even after the query was "fixed" to workaround the "infinite runtime" issue, the db server was still responding with the same "'Could not find index '214730889' in collection 's3010751'.' (while optimizing plan)" until the db server node was restarted. At that point ArangoDB resumed normal operation.

So to summarize three issues here:

1) Spurious error in the logs related to a missing index that is not missing at all, and

2) A faulty execution plan for a graph traversal query resulting in seemingly infinite runtime, and the spurious error already mentioned.

3) Somehow this (or something else?) corrupted the db server, causing the same error even on other queries, until the db server was restarted.

**Comment by Simran-B** [comment]
@rocketraman Thanks for filing an issue. I added the issue template to your post. Can you please update it with more information about your environment, configuration and query?

**Comment by rocketraman** [comment]
> @rocketraman Thanks for filing an issue. I added the issue template to your post. Can you please update it with more information about your environment, configuration and query?

@Simran-B Thanks, I've updated the OP and filled in the issue template.

**Comment by rocketraman** [comment]
I had this problem again today. The other thing that worked was deleting the index in question, which prevented the error from occurring. Also, possibly related: I tried to recreate the index after deleting it, and I see this error in the logs:

```
arangodb-db-2 arangodb 2020-04-01T02:35:39.978284712Z 2020-04-01T02:35:39Z [1] ERROR [bc555] {maintenance} EnsureIndex: failed to ensure index {"collection":"s3010751","deduplicate":false,"fields":["uuid"],"id":"277389673","inBackground":false,"isBuilding":true,"name":"idx_uuid","sparse":false,"type":"persistent","unique":true} unique constraint violated - in index idx_uuid of type persistent over 'uuid'; conflicting key: 159844632
```

However, there *seems* to be only one document with that key:

```
> db.files.document("159844632")
{
  "_key" : "159844632", 
  "_id" : "files/159844632", 
  "_rev" : "_Z7qTcfu--_",
  ...
}
```

I then tried to recreate the index *without* unique=true, which worked successfully, and executed a query to determine which entries were unique, and I do see two results with the same `_key` and `_id`, but with different `_rev`s:

```
> db._query("FOR f IN files COLLECT u = f.uuid INTO dups FILTER LENGTH(dups) > 1 RETURN { u, dups }");
[object ArangoQueryCursor, count: 1, cached: false, hasMore: false]

[ 
  { 
    "u" : "1yhPlir-R9y97ph6Kddr5Q", 
    "dups" : [ 
      { 
        "f" : { 
          "_key" : "159844632", 
          "_id" : "files/159844632", 
          "_rev" : "_ZWvGL3e--_", 
          ...
        } 
      }, 
      { 
        "f" : { 
          "_key" : "159844632", 
          "_id" : "files/159844632", 
          "_rev" : "_Z7qTcfu--_", 
          ...
        } 
      } 
    ] 
  } 
]
```

Is it possible that this data corruption is the underlying cause of the weird error `1590 Could not find index` error? How would this data corruption have occurred in the first place?

**Comment by rocketraman** [comment]
Also, I was able to delete one of the two entries, but the other one seems be some kind of zombie:

```
> db._query("FOR f IN files FILTER f.uuid == '1yhPlir-R9y97ph6Kddr5Q' RETURN f")
[object ArangoQueryCursor, count: 1, cached: false, hasMore: false]

[ 
  { 
    "_key" : "159844632", 
    "_id" : "files/159844632", 
    "_rev" : "_ZWvGL3e--_", 
    ...
  } 
]

> db._query("FOR f IN files FILTER f.uuid == '1yhPlir-R9y97ph6Kddr5Q' REMOVE f IN files")

JavaScript exception in file '/usr/share/arangodb3/js/client/modules/@arangodb/arangosh.js' at 99,7: ArangoError 1202: AQL: Error message received from cluster node 'PRMR-e6e30ee8-02f5-4707-8361-89cbc0f69c69': document not found (while executing)
!      throw error;
!      ^
stacktrace: ArangoError: AQL: Error message received from cluster node 'PRMR-e6e30ee8-02f5-4707-8361-89cbc0f69c69': document not found (while executing)
    at Object.exports.checkRequestResult (/usr/share/arangodb3/js/client/modules/@arangodb/arangosh.js:97:21)
    at ArangoStatement.execute (/usr/share/arangodb3/js/client/modules/@arangodb/arango-statement.js:172:12)
    at Proxy.ArangoDatabase._query (/usr/share/arangodb3/js/client/modules/@arangodb/arango-database.js:954:45)
    at <shell command>:1:4
```

**Comment by fceller** [comment]
Thanks @rocketraman for the additional info. We will try to create a test setup to reproduce the problem.
